File: apps/briefcase-mobile-signer/ios/BriefcaseMobileSigner/Core/NoiseNNpsk0.swift
import CryptoKit
import Foundation

// Minimal Noise implementation for a single handshake pattern:
//
//   Noise_NNpsk0_25519_ChaChaPoly_SHA256
//
// This is intentionally not a generic Noise framework. It exists so the mobile
// signer can pair with `briefcased` without shipping extra dependencies.

enum NoiseNNpsk0Error: Error {
    case invalidPskLength
    case invalidMessage
    case cryptoFailure
}

private func sha256(_ data: Data) -> Data {
    Data(SHA256.hash(data: data))
}

private func hmacSha256(key: Data, data: Data) -> Data {
    let symKey = SymmetricKey(data: key)
    let mac = HMAC<SHA256>.authenticationCode(for: data, using: symKey)
    return Data(mac)
}

private func hkdf(
    chainingKey: Data,
    inputKeyMaterial: Data,
    numOutputs: Int
) -> [Data] {
    // HKDF from the Noise spec (RFC5869 with salt=chainingKey, info empty).
    let tempKey = hmacSha256(key: chainingKey, data: inputKeyMaterial)
    let output1 = hmacSha256(key: tempKey, data: Data([0x01]))
    let output2 = hmacSha256(key: tempKey, data: output1 + Data([0x02]))
    if numOutputs == 2 {
        return [output1, output2]
    }
    let output3 = hmacSha256(key: tempKey, data: output2 + Data([0x03]))
    return [output1, output2, output3]
}

private struct CipherState {
    // If keyBytes is nil, encryption is a no-op (plaintext).
    private var keyBytes: Data?
    private var nonce: UInt64 = 0

    mutating func initializeKey(_ key: Data?) {
        keyBytes = key
        nonce = 0
    }

    private func nonceBytes() throws -> ChaChaPoly.Nonce {
        // Noise nonce: 96-bit nonce = 4 zero bytes || little_endian(nonce64)
        var n = nonce.littleEndian
        var out = Data(repeating: 0, count: 4)
        out.append(Data(bytes: &n, count: 8))
        return try ChaChaPoly.Nonce(data: out)
    }

    mutating func encryptWithAd(ad: Data, plaintext: Data) throws -> Data {
        guard let keyBytes else {
            return plaintext
        }
        let key = SymmetricKey(data: keyBytes)
        let n = try nonceBytes()
        let box = try ChaChaPoly.seal(plaintext, using: key, nonce: n, authenticating: ad)
        nonce &+= 1
        return box.ciphertext + box.tag
    }

    mutating func decryptWithAd(ad: Data, ciphertext: Data) throws -> Data {
        guard let keyBytes else {
            return ciphertext
        }
        guard ciphertext.count >= 16 else {
            throw NoiseNNpsk0Error.invalidMessage
        }
        let key = SymmetricKey(data: keyBytes)
        let n = try nonceBytes()
        let ct = ciphertext.prefix(ciphertext.count - 16)
        let tag = ciphertext.suffix(16)
        let box = try ChaChaPoly.SealedBox(nonce: n, ciphertext: ct, tag: tag)
        let pt = try ChaChaPoly.open(box, using: key, authenticating: ad)
        nonce &+= 1
        return pt
    }
}

private struct SymmetricState {
    private(set) var ck: Data
    private(set) var h: Data
    private var cipher: CipherState

    init(protocolName: String) {
        let pn = Data(protocolName.utf8)
        if pn.count <= 32 {
            var padded = pn
            padded.append(Data(repeating: 0, count: 32 - pn.count))
            h = padded
        } else {
            h = sha256(pn)
        }
        ck = h
        cipher = CipherState()
        cipher.initializeKey(nil)
    }

    mutating func mixHash(_ data: Data) {
        h = sha256(h + data)
    }

    mutating func mixKey(_ inputKeyMaterial: Data) {
        let outs = hkdf(chainingKey: ck, inputKeyMaterial: inputKeyMaterial, numOutputs: 2)
        ck = outs[0]
        cipher.initializeKey(outs[1])
    }

    mutating func mixKeyAndHash(_ inputKeyMaterial: Data) {
        let outs = hkdf(chainingKey: ck, inputKeyMaterial: inputKeyMaterial, numOutputs: 3)
        ck = outs[0]
        mixHash(outs[1])
        cipher.initializeKey(outs[2])
    }

    mutating func encryptAndHash(_ plaintext: Data) throws -> Data {
        let ciphertext = try cipher.encryptWithAd(ad: h, plaintext: plaintext)
        mixHash(ciphertext)
        return ciphertext
    }

    mutating func decryptAndHash(_ ciphertext: Data) throws -> Data {
        let plaintext = try cipher.decryptWithAd(ad: h, ciphertext: ciphertext)
        mixHash(ciphertext)
        return plaintext
    }
}

struct NoiseNNpsk0Initiator {
    private var symmetric = SymmetricState(protocolName: "Noise_NNpsk0_25519_ChaChaPoly_SHA256")
    private let psk: Data
    private var e: Curve25519.KeyAgreement.PrivateKey?

    init(psk: Data) throws {
        guard psk.count == 32 else {
            throw NoiseNNpsk0Error.invalidPskLength
        }
        self.psk = psk
    }

    mutating func writeMessage1(payload: Data = Data()) throws -> Data {
        // message1 pattern: psk, e
        symmetric.mixKeyAndHash(psk)

        let eph = Curve25519.KeyAgreement.PrivateKey()
        e = eph
        let ePub = eph.publicKey.rawRepresentation
        symmetric.mixHash(ePub)
        // PSK mode: MixKey(e.public_key) after MixHash(e.public_key).
        symmetric.mixKey(ePub)

        let ciphertext = try symmetric.encryptAndHash(payload)
        return ePub + ciphertext
    }

    mutating func readMessage2(_ message: Data) throws -> Data {
        // message2 pattern: e, ee
        guard message.count >= 32 + 16 else {
            throw NoiseNNpsk0Error.invalidMessage
        }
        guard let eph = e else {
            throw NoiseNNpsk0Error.invalidMessage
        }

        let rePub = message.prefix(32)
        let rest = message.dropFirst(32)

        symmetric.mixHash(rePub)
        symmetric.mixKey(rePub) // PSK mode.

        let re = try Curve25519.KeyAgreement.PublicKey(rawRepresentation: rePub)
        let ss = try eph.sharedSecretFromKeyAgreement(with: re)
        let dh = ss.withUnsafeBytes { Data($0) }
        symmetric.mixKey(dh)

        return try symmetric.decryptAndHash(Data(rest))
    }
}

File: apps/briefcase-mobile-signer/android/app/src/main/java/com/credentialbriefcase/mobilesigner/NoiseNNpsk0.kt
package com.credentialbriefcase.mobilesigner

import java.security.MessageDigest
import java.security.SecureRandom
import javax.crypto.Mac
import javax.crypto.spec.SecretKeySpec
import org.bouncycastle.crypto.InvalidCipherTextException
import org.bouncycastle.crypto.agreement.X25519Agreement
import org.bouncycastle.crypto.modes.ChaCha20Poly1305
import org.bouncycastle.crypto.params.AEADParameters
import org.bouncycastle.crypto.params.KeyParameter
import org.bouncycastle.crypto.params.X25519PrivateKeyParameters
import org.bouncycastle.crypto.params.X25519PublicKeyParameters

// Minimal Noise implementation for a single handshake pattern:
//
//   Noise_NNpsk0_25519_ChaChaPoly_SHA256
//
// This matches `docs/PAIRING.md` and `briefcased`'s `snow` usage.
class NoiseNNpsk0Initiator(psk: ByteArray) {
    private val psk: ByteArray
    private var symmetric = SymmetricState("Noise_NNpsk0_25519_ChaChaPoly_SHA256")
    private var e: X25519PrivateKeyParameters? = null

    init {
        require(psk.size == 32) { "psk must be 32 bytes" }
        this.psk = psk.copyOf()
    }

    fun writeMessage1(payload: ByteArray = ByteArray(0)): ByteArray {
        // message1: psk, e
        symmetric.mixKeyAndHash(psk)

        val rnd = SecureRandom()
        val eph = X25519PrivateKeyParameters(rnd)
        e = eph
        val ePub = eph.generatePublicKey().encoded
        symmetric.mixHash(ePub)
        // PSK mode: MixKey(e.public_key) after MixHash(e.public_key)
        symmetric.mixKey(ePub)

        val ciphertext = symmetric.encryptAndHash(payload)
        return ePub + ciphertext
    }

    fun readMessage2(message: ByteArray): ByteArray {
        // message2: e, ee
        if (message.size < 32 + 16) throw IllegalArgumentException("invalid message2")
        val eph = e ?: throw IllegalStateException("message1 not written")

        val rePubBytes = message.copyOfRange(0, 32)
        val rest = message.copyOfRange(32, message.size)

        symmetric.mixHash(rePubBytes)
        symmetric.mixKey(rePubBytes) // PSK mode.

        val re = X25519PublicKeyParameters(rePubBytes, 0)
        val dh = ByteArray(32)
        val agree = X25519Agreement()
        agree.init(eph)
        agree.calculateAgreement(re, dh, 0)
        symmetric.mixKey(dh)

        return symmetric.decryptAndHash(rest)
    }

    private class CipherState {
        private var key: ByteArray? = null
        private var nonce: Long = 0L

        fun initializeKey(keyBytes: ByteArray?) {
            key = keyBytes?.copyOf()
            nonce = 0L
        }

        private fun nonceBytes(): ByteArray {
            // 96-bit nonce = 4 zero bytes || little_endian(nonce64)
            val out = ByteArray(12)
            var n = nonce
            for (i in 0 until 8) {
                out[4 + i] = (n and 0xff).toByte()
                n = n ushr 8
            }
            return out
        }

        fun encryptWithAd(ad: ByteArray, plaintext: ByteArray): ByteArray {
            val k = key ?: return plaintext
            val nonceBytes = nonceBytes()

            val aead = ChaCha20Poly1305()
            aead.init(true, AEADParameters(KeyParameter(k), 128, nonceBytes, ad))

            val out = ByteArray(aead.getOutputSize(plaintext.size))
            var len = aead.processBytes(plaintext, 0, plaintext.size, out, 0)
            len += aead.doFinal(out, len)
            nonce += 1
            return out.copyOf(len)
        }

        fun decryptWithAd(ad: ByteArray, ciphertext: ByteArray): ByteArray {
            val k = key ?: return ciphertext
            val nonceBytes = nonceBytes()

            val aead = ChaCha20Poly1305()
            aead.init(false, AEADParameters(KeyParameter(k), 128, nonceBytes, ad))

            val out = ByteArray(aead.getOutputSize(ciphertext.size))
            try {
                var len = aead.processBytes(ciphertext, 0, ciphertext.size, out, 0)
                len += aead.doFinal(out, len)
                nonce += 1
                return out.copyOf(len)
            } catch (e: InvalidCipherTextException) {
                throw IllegalArgumentException("decrypt failed", e)
            }
        }
    }

    private class SymmetricState(protocolName: String) {
        private var cipher = CipherState()
        private var ck: ByteArray
        private var h: ByteArray

        init {
            val pn = protocolName.toByteArray(Charsets.UTF_8)
            h = if (pn.size <= 32) {
                pn + ByteArray(32 - pn.size)
            } else {
                sha256(pn)
            }
            ck = h.copyOf()
            cipher.initializeKey(null)
        }

        fun mixHash(data: ByteArray) {
            h = sha256(h + data)
        }

        fun mixKey(ikm: ByteArray) {
            val outs = hkdf(ck, ikm, 2)
            ck = outs[0]
            cipher.initializeKey(outs[1])
        }

        fun mixKeyAndHash(ikm: ByteArray) {
            val outs = hkdf(ck, ikm, 3)
            ck = outs[0]
            mixHash(outs[1])
            cipher.initializeKey(outs[2])
        }

        fun encryptAndHash(plaintext: ByteArray): ByteArray {
            val ciphertext = cipher.encryptWithAd(h, plaintext)
            mixHash(ciphertext)
            return ciphertext
        }

        fun decryptAndHash(ciphertext: ByteArray): ByteArray {
            val plaintext = cipher.decryptWithAd(h, ciphertext)
            mixHash(ciphertext)
            return plaintext
        }

        companion object {
            private fun sha256(data: ByteArray): ByteArray {
                val md = MessageDigest.getInstance("SHA-256")
                return md.digest(data)
            }

            private fun hmacSha256(key: ByteArray, data: ByteArray): ByteArray {
                val mac = Mac.getInstance("HmacSHA256")
                mac.init(SecretKeySpec(key, "HmacSHA256"))
                return mac.doFinal(data)
            }

            private fun hkdf(chainingKey: ByteArray, ikm: ByteArray, n: Int): List<ByteArray> {
                val tempKey = hmacSha256(chainingKey, ikm)
                val output1 = hmacSha256(tempKey, byteArrayOf(0x01))
                val output2 = hmacSha256(tempKey, output1 + byteArrayOf(0x02))
                if (n == 2) return listOf(output1, output2)
                val output3 = hmacSha256(tempKey, output2 + byteArrayOf(0x03))
                return listOf(output1, output2, output3)
            }
        }
    }
}

